### PR TITLE
[DO NOT MERGE] [DO NOT REVIEW][WIP] drivers: imx8m: SDMA controller driver

### PR DESCRIPTION
**Pull request by paulstelian97**
This patch set contains the i.MX SDMA controller driver as well as adding its support on the i.MX8M platform.

The driver is incomplete. The most important missing feature is adding a proper handshake implementation, so that the hardware DMA request number and SAI watermark are properly passed to the DMA controller and its builtin scripts.

There are also debug hacks on this branch which will disappear before I will mark this as ready for review. They should be mostly out of the way anyway.

Other missing parts can be discovered by looking at the various TODOs sprinkled in the code.

### CODE REVIEW

**Comment by dbaluta** [comment]
Thanks @paulstelian97 awesome effort on your side! 

Cc: @TE-N-ShengjiuWang @tanglez for review.

**Comment by dbaluta** [comment]
Fixes: https://github.com/thesofproject/sof/pull/2483

**Comment by dbaluta** [comment]
Fixes: https://github.com/thesofproject/sof/issues/2330

File: src/drivers/imx/sdma.c [hunk 1]
```diff
@@ -0,0 +1,716 @@
...
+		}
+	}
+	/* Checks passed, we need to populate the buffer descriptors,
+	 * adjust count, configure context so the FIFO address is also
+	 * stored, upload context and have everything be ready.
+	 */
+
+	/* TODO fetch HW event number and proper watermark */
+	int hwevent = -1;
+	int watermark = 16;
```
Comment (lgirdwood) [inline src/drivers/imx/sdma.c]: Magic numbers.

File: src/drivers/imx/sdma.c [hunk 2]
```diff
@@ -0,0 +1,716 @@
...
+		if (!src_may_change)
+			pdata->descriptors[i].buf_addr = config->elem_array.elems[i].dest;
+		if (!dst_may_change)
+			pdata->descriptors[i].buf_addr = config->elem_array.elems[i].src;
+		if (src_may_change && dst_may_change) {
+			/* M2M copy */
+			pdata->descriptors[i].buf_addr = config->elem_array.elems[i].src;
+			pdata->descriptors[i].buf_xaddr = config->elem_array.elems[i].dest;
+		}
+		int width = 32;
```
Comment (lgirdwood) [inline src/drivers/imx/sdma.c]: Looks like we are hard coding the format here too ? Magic number.

**Comment by dbaluta** [comment]
@lgirdwood thanks for taking a look. Lets do a serious review next week.

I think @paulstelian97 only pushed this PR for me to have a look and test.

File: src/drivers/imx/sdma.c (hunk 1 above)
Comment (paulstelian97) [reply src/drivers/imx/sdma.c]: Forgot to force push, these are already gone on my side. hwevent will come from handshake, watermark comes from dma config burst_elems.

File: src/drivers/imx/sdma.c (hunk 2 above)
Comment (paulstelian97) [reply src/drivers/imx/sdma.c]: It's fetched from src_width or dst_width accordingly, forgot to force push.

File: src/drivers/imx/sdma.c [hunk 3]
```diff
@@ -95,10 +97,13 @@ static int sdma_run_c0(struct dma *dma, uint8_t cmd, uint32_t buf_addr, uint16_t
 	/* 1 is BIT(0) for channel 0, the bit will be cleared as the
 	 * channel finishes
 	 */
-	ret = poll_for_register_delay(dma_base(dma) + SDMA_STOP_STAT, 1, 0, 500);
+	ret = poll_for_register_delay(dma_base(dma) + SDMA_STOP_STAT, 1, 0, 400000);
```
Comment (paulstelian97) [inline src/drivers/imx/sdma.c]: (I have to figure out a good polling interval that isn't huge but is enough for the actual transfers to work; perhaps one milisecond suffices but this needs testing)

**Comment by paulstelian97** [comment]
Closing for now, will resubmit once #2575 is merged.

**Comment by paulstelian97** [comment]
Replaced by #2595.

**Pull request closed without merging**